### PR TITLE
Add 'Emoncms in' to fetch latest feed value from emoncms server

### PR DESCRIPTION
**Pull request by glynhudson**
First pull request, please let me know if I've missed anything!

### CODE REVIEW

**Comment by hardillb** [comment]
New nodes don't need to be added to the node-red-nodes group these days, you should package them as npm modules following the guide (http://nodered.org/docs/creating-nodes/packaging.html) and publish them yourself.

Also we can't accept any code without a signed CLA as mentioned in the contribution guidelines. (https://github.com/node-red/node-red/blob/master/CONTRIBUTING.md)

**Comment by dceejay** [comment]
Ben,

As its adding to a node (type) we already have then adding it here would be appropriate. But does require a CLA as you point out.

**Comment by glynhudson** [comment]
HI Ben / Dave, 

Thanks for your guidance. I have emailed a completed and signed CLA to Nick this afternoon. 
Yes, this is an update to an existing module. Is there anything else I need to do to help you accept my pull request?  I notice the automated check also threw a 'Travis CI build failed' error. 

Thanks,

**Comment by knolleary** [comment]
node-red-gitbot: please recheck cla status

**Comment by knolleary** [comment]
Travis errors are ongoing known issues we haven't sorted out - not going to hold them against you.

You have changed the type of an existing node - that will break users, so we cannot do that. The `emoncms` node must keep its name.

If you can undo that bit of the PR, will be happy to merge this in.

**Comment by glynhudson** [comment]
Do you mean like this, I've changed the name of the old node back: https://github.com/glynhudson/node-red-nodes/commit/1467575f6c86d59f0dab202ce41a5ae5c44f0c12

Let me know, I'll re submit PR

File: io/emoncms/88-emoncms.js [hunk 1]
```diff
@@ -80,4 +80,52 @@ module.exports = function(RED) {
...
+            var feedid = this.feedid || msg.feedid;
+            if (feedid !== "") {
+                this.url += '&id=' + feedid;
+            }
+            if (typeof msg.time !== 'undefined') {
+                this.url += '&time=' + msg.time;
+            }
+            node.log("[emoncms] "+this.url);
+            http.get(this.url, function(res) {
+                node.log("Http response: " + res.statusCode);
```
Comment (knolleary) [inline io/emoncms/88-emoncms.js]: We don't generally log 'success' like this - if it's operating normally then it ought to be quiet

File: io/emoncms/88-emoncms.js [hunk 2]
```diff
@@ -80,4 +80,52 @@ module.exports = function(RED) {
...
+                    node.send(msg);
+                }
+                res.setEncoding('utf8');
+                res.on('data', function(chunk) {
+                    msg.payload += chunk;
+                });
+                res.on('end', function() {
+                    node.send(msg);
+                });
+            }).on('error', function(e) {
```
Comment (knolleary) [inline io/emoncms/88-emoncms.js]: If there's an error, we shouldn't pass it on - use node.error to log the condition.

File: io/emoncms/88-emoncms.js [hunk 3]
```diff
@@ -80,4 +80,52 @@ module.exports = function(RED) {
...
+                msg.payload = "";
+                if ((msg.rc != 200) && (msg.rc != 404)) {
+                    node.send(msg);
+                }
+                res.setEncoding('utf8');
+                res.on('data', function(chunk) {
+                    msg.payload += chunk;
+                });
+                res.on('end', function() {
+                    node.send(msg);
```
Comment (knolleary) [inline io/emoncms/88-emoncms.js]: Does the emoncms api always return JSON here? If so, we ought to parse the body for the user and pass that on.

**Comment by glynhudson** [comment]
Hi guys, think I'm now good to go with this one. I have taken on board what you said about not logging success and phasing JSON before it's passed on and made changes accordingly. Please can you review and merge. Thanks a lot :-)

**Comment by glynhudson** [comment]
Any update? Have you had a chance to review and merge? Thanks a lot

**Comment by knolleary** [comment]
On my list to look at today/tomorrow.

**Comment by glynhudson** [comment]
Sorry to hassle, any chance you could review and merge? I'm working on an
article using the emoncms node red module and really need these changes to
be merged before I publish. One thing I wasn't sure about is the json
output  being correctly phased? I still seem to need the json processing
module to get rid of the ""

- sent from my mobile device

**Comment by knolleary** [comment]
Top of my list once I fix the current issue I'm on. Expect a merge (or comments) in the next hour.

File: io/emoncms/88-emoncms.js [hunk 4]
```diff
@@ -75,9 +73,50 @@ module.exports = function(RED) {
...
+            this.url += '&apikey='+this.apikey;
+            var feedid = this.feedid || msg.feedid;
+            if (feedid !== "") {
+                this.url += '&id=' + feedid;
+            }
+            http.get(this.url, function(res) {
+                msg.rc = res.statusCode;
+                msg.payload = "";
+                if ((msg.rc != 200) && (msg.rc != 404)) {
+                    node.send(JSON.parse(msg));
```
Comment (knolleary) [inline io/emoncms/88-emoncms.js]: This needs moving down to line 112, inside the on(end) handler to replace the node.send call already there.

**Comment by knolleary** [comment]
One minor change needed - but given we've left you waiting for this, I'll merge it and make the change myself.

**Comment by knolleary** [comment]
@glynhudson I've just pushed a slight tweak to your change - you were parsing the JSON response in the error handling path, not the success path. Can you grab the latest file and see if it works as you expect? If you're happy with it, I'll publish this version to npm.

**Comment by glynhudson** [comment]
Great, thanks a lot. Wops, that was a silly mistake. Sorry I'm new to nodejs. Thanks a lot for correcting and merging for me. Please could you publish to NPM. I have published my article: http://openenergymonitor.blogspot.com/2015/11/ambient-wind-energy-indicator-using.html

**Comment by knolleary** [comment]
node-red-node-emoncms@0.0.5 published

**Pull request merged**